lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp

```diff
@@ -214,4 +223,112 @@ void LLDBMemoryReader::popLocalBuffer() {
+  // doing memory reads later on. This should not clash with any masks used by
+  // Swift.
+  uint64_t next_available_address = 0x2000000000000000;
+  if (!m_address_map.empty()) {
+    // We map the images contiguously one after the other, all with the highest
+    // bit set.
+    next_available_address = m_address_map.back().first;
+    // Add 8 to it to keep it pointer aligned. This assumes the previous address
+    // was also pointer aligned.
+    next_available_address = next_available_address + 8;
+  }
+
+#ifndef NDEBUG
+  static std::initializer_list<uint64_t> used_swift_masks = {
+      SWIFT_ABI_ARM_OBJC_WEAK_REFERENCE_MARKER_MASK,
+      SWIFT_ABI_ARM_OBJC_WEAK_REFERENCE_MARKER_VALUE,
+      SWIFT_ABI_ARM_IS_OBJC_BIT,
+      SWIFT_ABI_X86_64_SWIFT_SPARE_BITS_MASK,
+      SWIFT_ABI_X86_64_OBJC_RESERVED_BITS_MASK,
+      SWIFT_ABI_X86_64_SIMULATOR_OBJC_RESERVED_BITS_MASK,
+      SWIFT_ABI_X86_64_IS_OBJC_BIT,
+      SWIFT_ABI_X86_64_OBJC_WEAK_REFERENCE_MARKER_MASK,
+      SWIFT_ABI_X86_64_SIMULATOR_OBJC_WEAK_REFERENCE_MARKER_MASK,
+      SWIFT_ABI_X86_64_SIMULATOR_OBJC_WEAK_REFERENCE_MARKER_VALUE,
+      SWIFT_ABI_ARM64_SWIFT_SPARE_BITS_MASK,
+      SWIFT_ABI_ARM64_OBJC_RESERVED_BITS_MASK,
+      SWIFT_ABI_ARM64_IS_OBJC_BIT,
+      SWIFT_ABI_ARM64_OBJC_WEAK_REFERENCE_MARKER_MASK};
+
+  for (auto swift_mask : used_swift_masks) {
+    assert((next_available_address & swift_mask) != swift_mask &&
+           "LLDB artificial address clashes with a swift mask!");
+  }
+#endif
+
+  SectionList *section_list = module->GetObjectFile()->GetSectionList();
+  auto last_section =
+      section_list->GetSectionAtIndex(section_list->GetSize() - 1);
+  // The virtual file address + the size of last section gives us the total size
+  // of this image in memory.
+  uint64_t size = last_section->GetFileAddress() + last_section->GetByteSize();
+  auto next_available_address_end = next_available_address + size;
+  next_available_address_end = next_available_address_end + 7 & ~7;
+  m_address_map.emplace_back(next_available_address_end, module);
+  return {next_available_address, next_available_address_end};
+}
+
+llvm::Optional<Address>
+LLDBMemoryReader::tryToConvertFromMappedAddressIntoRealAddress(
+    uint64_t address) const {
+  Log *log(GetLogIfAllCategoriesSet(LIBLLDB_LOG_TYPES));
+
+  if (!m_process.GetTarget().GetSwiftReadMetadataFromFileCache())
+    return {};
+
+  // If the address contains our mask, this is an image we registered.
+  if (address & 0x2000000000000000) {
+    // Pair for the lower_bound function. Pair compares the first element first,
+    // so it should only ever need to compare the addresses and never the
+    // module.
+    auto comparison_pair = std::make_pair(address, ModuleSP());
+    auto pair_iterator = std::lower_bound(m_address_map.begin(),
+                                          m_address_map.end(), comparison_pair);
+    if (pair_iterator != m_address_map.end()) {
```

Comment:
std::lowerbound could return an entry that doesn't contain the searched address, since it returns the insert point for that value. Can you add error handling for that?

Reply:
Sorry, I don't understand, in what situation can that happen?